src/main/java/edu/harvard/dbmi/avillach/dictionaryetl/facet/RecoverMonthsFacetGeneratorService.java

```diff
@@ -0,0 +1,273 @@
+                        parentName, recoverAdultFacetModel.getDisplay(), recoverAdultFacetModel.getDescription(),
+                        studyId, adultNodeRegex, months
+                );
+
+                if (wrapper == null || wrapper.facetCategory == null || wrapper.facetCategory.facets == null || wrapper.facetCategory.facets.isEmpty()) {
+                    out.message = "No months discovered; nothing to load.";
+                    return out;
+                }
+
+                out.load = facetLoaderService.load(List.of(wrapper));
+                out.message = "Generation complete.";
+            } else {
+                out.message = "Recover adult facet metadata is missing.";
+            }
+
+        } else {
+            boolean consortiumFacetMissing = consortiumCuratedFacets.isEmpty();
+            boolean recoverAdultFacetMissing = recoverAdultOptional.isEmpty();
+
+            if (consortiumFacetMissing && recoverAdultFacetMissing) {
+                out.message = "Consortium Curated facet and Recover adult facet is missing.";
+            } else if (consortiumFacetMissing) {
+                out.message = "Consortium Curated facets is missing.";
+            } else {
+                out.message = "Recover Adult facet is missing.";
+            }
+        }
+
+        return out;
+    }
+
+    private Set<Integer> discoverMonths(String studyId, String pathPrefixRegex) {
+        Set<Integer> months = new TreeSet<>();
+        final Pattern compiledPrefix = compilePrefix(pathPrefixRegex);
+
+        try (Stream<ConceptPathRow> rows = conceptRepository.streamDatasetNodeIdAndPath(studyId)) {
+            rows.forEach(row -> {
+                String path = row.getConceptPath();
+                if (compiledPrefix != null && (path == null || !compiledPrefix.matcher(path).find())) return;
+
+                List<String> nodes = split(path);
+                int n = nodes.size();
+                if (n == 0) return;
+
+                String last = nodes.get(n - 1);
+                String prev = n >= 2 ? nodes.get(n - 2) : null;
+
+                // Case 1 & 2: previous node = (Inf|Noninf) and last node = integer or last node = minus3
+                if (prev != null && last != null
+                    && prev.matches("(?i)^(inf|infected|noninf|noninfected)$")) {
+                    if(INT_PATTERN.matcher(last).matches()) {
+                        try {
+                            months.add(Integer.parseInt(last));
+                            return;
+                        } catch (NumberFormatException ignored) {
+                            // ignore non-numeric tails
+                        }
+                    }
+
+                    Matcher minusMatcher = MINUS_3_PATTERN.matcher(last);
+                    if (minusMatcher.matches()) {
+                        months.add(-Integer.parseInt(minusMatcher.group(1)));
+                        return;
```

Comment:
just checking - if you match to the first pattern but not the second pattern, it's ok to match to the third, but if you match to the second pattern, it's not ok to match to the third? tldr that return seems like part of an incomplete thought.

Reply:
This is not an incomplete thought. The logic applies specifically to the RECOVER Adult concept paths in BDC.

If the second-to-last node matches the first pattern (?i)^(inf|infected|noninf|noninfected)$, then only the INT_PATTERN or MINUS_3_PATTERN will be evaluated.

If it does not match the first pattern, the code checks whether the "months post index" value is embedded elsewhere in the concept path based on the suffix.